core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java

```diff
@@ -408,19 +425,27 @@ public boolean equals(Object o) {
         if (!customs.equals(that.customs)) {
             return false;
         }
+
+        if (Arrays.equals(primaryTerms, that.primaryTerms) == false) {
+            return false;
+        }
         return true;
     }
 
     @Override
     public int hashCode() {
         int result = index.hashCode();
+        result = 31 * result + (int) (version ^ (version >>> 32));
```

Comment:
There's a built-in for this `[Long#hashCode(long)](https://docs.oracle.com/javase/8/docs/api/java/lang/Long.html#hashCode-long-)` since Java 8.